### PR TITLE
CaptainAgent Finalize

### PR DESCRIPTION
**Pull request by LeoLjl**
Added notebook, test cases.

## Checks

- [ ] I've included any doc changes needed for https://ag2ai.github.io/ag2/. See https://ag2ai.github.io/ag2/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Review by qingyun-wu — APPROVED** [review]
LGTM! Thanks Jiale!

**Comment by qingyun-wu** [comment]
@marklysze Could you take a look at the notebook content?

**Comment by marklysze** [comment]
> @marklysze Could you take a look at the notebook content?

Yes, will do

**Comment by qingyun-wu** [comment]
@LeoLjl, could you add an auto-test in the workflow file? https://github.com/ag2ai/ag2/blob/main/.github/workflows/contrib-openai.yml

File: notebook/agentchat_captainagent.ipynb [hunk 1]
```diff
@@ -0,0 +1,1926 @@
...
+    "# CaptainAgent\n",
+    "By: Jiale Liu, Linxin Song, Jieyu Zhang, Shaokun Zhang\n",
+    "\n",
+    "In this notebook, we introduce CaptainAgent, an agent enhanced with the capability to call AutoBuild to break down and solve complex tasks. AutoBuild can initiate a group chat between a group of experts and converse to solve the task. The experts in nested chat can be retrieved from agent library. The agents can be equipped with tools for advanced coding.\n",
+    "\n",
+    "````{=mdx}\n",
+    ":::info Requirements\n",
+    "Install `pyautogen`:\n",
+    "```bash\n",
+    "pip install pyautogen[autobuild]\n",
```
Comment (sonichi) [inline notebook/agentchat_captainagent.ipynb:16]: Does this section need updates?

File: notebook/agentchat_captainagent.ipynb [hunk 2]
```diff
@@ -0,0 +1,1926 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from autogen import UserProxyAgent\n",
+    "from autogen.agentchat.contrib.captainagent import CaptainAgent\n",
+    "\n",
+    "nested_config = {\n",
+    "    # this is used to configure the autobuild building process\n",
+    "    \"autobuild_build_config\": {\n",
+    "        # this is used to configure the user proxy within nested chat\n",
```
Comment (sonichi) [inline notebook/agentchat_captainagent.ipynb:548]: I don't get why "user proxy" is mentioned here. The line is about where to find the agent library.
Also, where would the file be expected to found?

File: notebook/agentchat_captainagent.ipynb [hunk 3]
```diff
@@ -0,0 +1,1926 @@
...
+    "        \"library_path_or_json\": \"captainagent_expert_library.json\",\n",
```
Comment (sonichi) [inline notebook/agentchat_captainagent.ipynb:549]:
```suggestion
    "        \"agent_library_path_or_json\": \"captainagent_expert_library.json\",\n",
```

File: notebook/agentchat_captainagent.ipynb [hunk 4]
```diff
@@ -0,0 +1,1926 @@
...
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from autogen import UserProxyAgent\n",
+    "from autogen.agentchat.contrib.captainagent import CaptainAgent\n",
+    "\n",
+    "nested_config = {\n",
+    "    # this is used to configure the autobuild building process\n",
+    "    \"autobuild_build_config\": {\n",
```
Comment (sonichi) [inline notebook/agentchat_captainagent.ipynb:547]: "autobuild_build_config" is hard to understand.
Let's discuss all the options and see if we can make them intuitive.

File: notebook/agentchat_captainagent.ipynb [hunk 5]
```diff
@@ -0,0 +1,1926 @@
...
+    "        \"library_path_or_json\": \"captainagent_expert_library.json\",\n",
+    "    },\n",
+    "    # this is used to configure the tool library\n",
+    "    \"autobuild_tool_config\": {\n",
```
Comment (sonichi) [inline notebook/agentchat_captainagent.ipynb:1432]: I'd like to see consistency in naming of the parameters. Not sure whether `autobuild_build_config` and `autobuild_tool_config` are good names.

File: notebook/agentchat_captainagent.ipynb (hunk 1 above)
Comment (LeoLjl) [reply notebook/agentchat_captainagent.ipynb:16]: Don't think so, it depends on Autobuild features.

File: notebook/agentchat_captainagent.ipynb (hunk 2 above)
Comment (LeoLjl) [reply notebook/agentchat_captainagent.ipynb:548]: Good catch, I removed the comment.

Comment (LeoLjl) [reply notebook/agentchat_captainagent.ipynb:548]: This agent library file is located in the same directory as the notebook.

File: notebook/agentchat_captainagent.ipynb (hunk 3 above)
Comment (LeoLjl) [reply notebook/agentchat_captainagent.ipynb:549]: This is not changeable at the moment, the arguments are passed exactly to AgentBuilder.build() [link](https://github.com/ag2ai/ag2/blob/90b90c6bf1ff471aca509423fba7c26926ab3b03/autogen/agentchat/contrib/agent_builder.py#L365), which takes in `library_path_or_json` as key.

Comment (LeoLjl) [reply notebook/agentchat_captainagent.ipynb:549]: We should keep it as it is now. And change this overall naming in the future PR.

File: autogen/agentchat/contrib/captainagent.py [hunk 6]
```diff
@@ -155,6 +157,8 @@ def __init__(
             max_consecutive_auto_reply (int): the maximum number of consecutive auto replies.
                 default to None (no limit provided, class attribute MAX_CONSECUTIVE_AUTO_REPLY will be used as the limit in this case).
                 The limit only plays a role when human_input_mode is not "ALWAYS".
+            agent_lib (str): the path to the agent library for retrieving the nested chat instantiated by CaptainAgent.
```
Comment (qingyun-wu) [inline autogen/agentchat/contrib/captainagent.py:160]: It could be a path or a json file

Comment (qingyun-wu) [inline autogen/agentchat/contrib/captainagent.py:160]:
```suggestion
            agent_lib (str): the path or a JSON file of the agent library for retrieving the nested chat instantiated by CaptainAgent.
```

**Pull request merged**